### ISSUE TITLE
Can some kind of logic be added to prevent unnecessary shipping method calculations?

### ISSUE THREAD

**Issue by WillBrubaker** [Automattic/woocommerce-services#2520]
Can this bit of code: https://github.com/Automattic/woocommerce-services/blob/1.25.21/woocommerce-services.php#L1095-L1128 be wrapped in conditional logic so it doesn't run if the order doesn't require shipping?

I've got an issue open for USPS here: https://github.com/woocommerce/woocommerce-shipping-usps/issues/344  The fatal is happening in USPS, but the order in question doesn't require shipping.

```
2022-02-01T00:12:30+00:00 CRITICAL Uncaught TypeError: sizeof(): Argument #1 ($value) must be of type Countable|array, null given in /nas/content/live/pisgahbanjos/wp-content/plugins/woocommerce-shipping-usps/includes/box-packer/class-wc-boxpack.php:114
Stack trace:
#0 /nas/content/live/pisgahbanjos/wp-content/plugins/woocommerce-shipping-usps/includes/class-wc-shipping-usps.php(1560): WC_Boxpack->pack()
#1 /nas/content/live/pisgahbanjos/wp-content/plugins/woocommerce-shipping-usps/includes/class-wc-shipping-usps.php(1159): WC_Shipping_USPS->box_shipping(Array)
#2 /nas/content/live/pisgahbanjos/wp-content/plugins/woocommerce-shipping-usps/includes/class-wc-shipping-usps.php(855): WC_Shipping_USPS->get_standard_package_requests(Array)
#3 /nas/content/live/pisgahbanjos/wp-content/plugins/woocommerce/includes/abstracts/abstract-wc-shipping-method.php(233): WC_Shipping_USPS->calculate_shipping(Array)
#4 /nas/content/live/pisgahbanjos/wp-content/plugins/woocommerce/includes/class-wc-shipping.php(348): WC_Shipping_Method->get_rates_for_package(Array)
#5 /nas/content/live/pisgahbanjos/wp-content/plugins/woocommerce-services/woocommerce-services.php(1099): WC_Shipping->calculate_shipping_for_package(Array)
#6 /nas/content/live/pisgahbanjos/wp-includes/class-wp-hook.php(307): WC_Connect_Loader->track_completed_order(4086, Array, Object(Automattic\WooCommerce\Admin\Overrides\Order))
#7 /nas/content/live/pisgahbanjos/wp-includes/class-wp-hook.php(331): WP_Hook->apply_filters(NULL, Array)
#8 /nas/content/live/pisgahbanjos/wp-includes/plugin.php(474): WP_Hook->do_action(Array)
#9 /nas/content/live/pisgahbanjos/wp-content/plugins/woocommerce/includes/class-wc-checkout.php(1183): do_action('woocommerce_che...', 4086, Array, Object(Automattic\WooCommerce\Admin\Overrides\Order))
#10 /nas/content/live/pisgahbanjos/wp-content/plugins/woocommerce/includes/class-wc-ajax.php(461): WC_Checkout->process_checkout()
#11 /nas/content/live/pisgahbanjos/wp-includes/class-wp-hook.php(307): WC_AJAX::checkout('')
#12 /nas/content/live/pisgahbanjos/wp-includes/class-wp-hook.php(331): WP_Hook->apply_filters('', Array)
#13 /nas/content/live/pisgahbanjos/wp-includes/plugin.php(474): WP_Hook->do_action(Array)
#14 /nas/content/live/pisgahbanjos/wp-content/plugins/woocommerce/includes/class-wc-ajax.php(90): do_action('wc_ajax_checkou...')
#15 /nas/content/live/pisgahbanjos/wp-includes/class-wp-hook.php(307): WC_AJAX::do_wc_ajax('')
#16 /nas/content/live/pisgahbanjos/wp-includes/class-wp-hook.php(331): WP_Hook->apply_filters(false, Array)
#17 /nas/content/live/pisgahbanjos/wp-includes/plugin.php(474): WP_Hook->do_action(Array)
#18 /nas/content/live/pisgahbanjos/wp-includes/template-loader.php(13): do_action('template_redire...')
#19 /nas/content/live/pisgahbanjos/wp-blog-header.php(19): require_once('/nas/content/li...')
#20 /nas/content/live/pisgahbanjos/index.php(17): require('/nas/content/li...')
#21 {main}
  thrown in /nas/content/live/pisgahbanjos/wp-content/plugins/woocommerce-shipping-usps/includes/box-packer/class-wc-boxpack.php on line 114
```

**Comment by waclawjacek** [comment]
Thanks for the report, @WillBrubaker! Is the fatal also thrown in WCS&T or is something else happening?

**Comment by WillBrubaker** [comment]
@waclawjacek I don't think the fatal is thrown in WCS&T

**Comment by waclawjacek** [comment]
I see, thanks, @WillBrubaker! For the person who starts working on this: could you please look into if the same thing that happens in the USPS extension also happens in WCS&T? Since this does not seem to happen in WCS&T, I'm assigning low priority to the issue.

**Comment by iyut** [comment]
Hi, I tried running some tests on WCS&T by creating an order with virtual product only. And after running a couple of tests, I don't see any fatal error nor any warning when I looked into debug.log.

I think one way to make sure if the order require shipping is by checking on `WC_Order::get_shipping_methods()` return value. if it does not return an empty array, then I believe the order is require shipping. Perhaps we can create a condition before the loop like below :
```
if ( 0 < count( $order->get_shipping_methods() ) ) {
  return;
}
```
What do you guys think?

**Comment by waclawjacek** [comment]
Hmm, I see there is already a call to `$packages = WC()->cart->get_shipping_packages()`. Does this function maybe also do what you mentioned? If so, we might not have to do anything after all - if it returns an empty array, the `foreach` will not execute.

**Comment by iyut** [comment]
I checked the `WC()->cart->get_shipping_packages()` function. It seems the function will always return an array based on this [line](https://woocommerce.github.io/code-reference/files/woocommerce-includes-class-wc-cart.html#source-view.1494). Unless someone desperately change the return value into another variable type using `woocommerce_cart_shipping_packages` filter ( which is IMO unlikely going to happen ).
I think you are right, we dont really need to do anything. Should we close this issue?

**Comment by berzley** [comment]
Experiencing this issue on a `virtual` product variation ( other variations require shipping ).

In the process of trying to troubleshoot out third-party plugins. Confirmed no in-house custom theme or plugins containing conflicting code.

#### Site Info:
```
WC Version: 6.2.1
REST API Version: ✔ 6.2.1
WC Blocks Version: ✔ 6.7.3
Action Scheduler Version: ✔ 3.4.0
WC Admin Version: ✔ 3.2.1
Log Directory Writable: ✔
WP Version: 5.9.1
WP Multisite: –
Advanced Custom Fields PRO: by Delicious Brains – 5.12
Akismet Anti-Spam: by Automattic – 4.2.2
Classic Editor: by WordPress Contributors – 1.6.2
GDPR Cookie Consent: by WebToffee – 2.1.1
Custom Facebook Feed Pro Smash: by Smash Balloon – 4.2.2
Custom Twitter Feeds Pro Developer: by Smash Balloon – 1.14.1
Formidable MailChimp: by Strategy11 – 2.06
Formidable Forms Pro: by Strategy11 – 5.2.01
Formidable Visual Views: by Strategy11 – 5.1.05
Formidable Forms: by Strategy11 – 5.2.01
Instagram Feed Pro Developer: by Smash Balloon – 6.0.4
Jetpack: by Automattic – 10.7
Smart Slider 3 Pro: by Nextend – 3.5.1.4
PrintAura WooCommerce API: by Print Aura – 4.0.2
Really Simple SSL: by Really Simple Plugins – 5.2.3
Redirection: by John Godley – 5.2.3
Rock The Slackbot: by Rachel Carden – 1.1.2
Smash Balloon Social Wall: by Smash Balloon – 1.0.6
Advanced Editor Tools (previously TinyMCE Advanced): by Automattic – 5.6.0
WooCommerce Admin: by WooCommerce – 3.2.1
WooCommerce Cart Add-Ons: by WooCommerce – 2.1.0
WooCommerce Stripe Gateway: by WooCommerce – 6.2.0
WooCommerce Shipping & Tax: by WooCommerce – 1.25.23
WooCommerce USPS Shipping: by WooCommerce – 4.4.69
WooCommerce: by Automattic – 6.2.1 (update to version 6.3.0 is available)
Wordfence Security: by Wordfence – 7.5.8
Yoast SEO: by Team Yoast – 18.2
WP Mail SMTP: by WPForms – 3.3.0
Feeds for YouTube Pro Developer: by Smash Balloon – 1.4
```

**Comment by iyut** [comment]
Hi @berzley , thank you for reporting this. Is there any step that I can follow to replicate the issue?

**Comment by berzley** [comment]
@iyut With the above environment I created a new product, checked "virtual" and "downloadable", and gave the product a sku value. Then adding the product to cart, proceeding to checkout, and then finally checking out. At checkout, after selecting payment option (paypal in my case) before payment processes the internal server error happens.

I repeated the steps again, but this time only selecting "virtual" and leaving all fields blank. Same results.

**Comment by iyut** [comment]
Hi @berzley , I've tried to replicate the issue on several PHP version ( 7.4, 8.0, 8.1 ). But still I seems cannot replicate the issue with both WC Shipping and Tax and WC Shipping USPS installed. Do you mind to test if the issue still remains if WC Shipping and Tax plugin is deactivated?

**Comment by berzley** [comment]
Hey @iyut, I found my solution.

tldr; downgrading PHP to 7.4 solved the issue

I did send a support message to WooCommerce support for the USPS shipping method and referenced this issue.

If I deactivate either WC Shipping and Tax or WC Shipping USPS (while leaving the other active or deactivating both) the virtual products work.

Additionally, I did a direct copy of the site and ran it in my local dev environment. Straight copy of the site, no changes, the smae products that fail on the live site work on the local testing site. Only difference is Apache w/ PHP 8.0 vs Nginx w/ PHP 7.4.

Tried the live site on 7.4, it worked. Live site back to 8.0 and it failed.

**Comment by berzley** [comment]
The issue is 100% in WC Shipping USPS. 
As of 8.0 `sizeof` or `count` throws a `TypeError` if the value isn't countable. Looks like `$this->items` never gets initialized in this case.

**Comment by iyut** [comment]
Thanks for the confirmation @berzley , since this is a WC Shipping USPS issue, should we move the discussion to [this USPS issue #344](https://github.com/woocommerce/woocommerce-shipping-usps/issues/344) ? Also, can you confirm if [this solution](https://github.com/woocommerce/woocommerce-shipping-usps/issues/344#issuecomment-1027736806) works for you?

**Comment by berzley** [comment]
@iyut Is it possible to get added to that repo? It's private.

**Comment by iyut** [comment]
Oh, sorry about that @berzley. There is a solution that might fix the issue in that USPS thread. 
Can you try to apply this condition below in `includes/box-packer/class-wc-boxpack.php` on line 114.

```
if ( null === $this->items || sizeof( $this->items ) == 0 ) {
	throw new Exception( 'No items to pack!' );
}
```
or probably this one is simpler : 
```
if ( empty( $this->items ) ) {
	throw new Exception( 'No items to pack!' );
}
```
See if one of those will fix the error issue.

**Comment by berzley** [comment]
Both of those work.

Also this would work
```php
if ( !is_array( $this->items ) ) {
	throw new Exception( 'No items to pack!' );
}
```

I believe initializing `$this->items` would work too
```php
private $items = array();
```

**Comment by csmcneill** [comment]
4859962-zen

**Comment by iyut** [comment]
Hi @berzley , USPS Shipping method version 4.4.70 has been released. It includes the fixes for `sizeof()` issue on PHp 8.0.15. Can you try to update your plugin and confirm if it fixes the issue?

**Comment by berzley** [comment]
@iyut Tests passed great and the client confirms the features are all working for them again with the this update.
Thanks!

**Comment by iyut** [comment]
Thanks for the confirmation @berzley ! 

Since the issue is from WC Shipping USPS ( [confirmed](https://github.com/Automattic/woocommerce-services/issues/2520#issuecomment-1068446946) by @berzley ) and we will not do anything on WC Services for this issue ( at least for now ), I will close this issue. Please feel free to re-open the issue if someone encounter this issue on WC Services.
